asab/storage/elasticsearch.py

```diff
@@ -93,49 +160,96 @@ async def get(self, index: str, obj_id: str, decrypt=None) -> dict:
 		"""
 		if decrypt is not None:
 			raise NotImplementedError("AES encryption for ElasticSearch not implemented")
-
 		for url in self.ServerUrls:
-			url = "{}{}/_doc/{}".format(url, index, obj_id)
+			request_url = "{}{}/_doc/{}".format(url, index, obj_id)
 			try:
-				async with self.session().request(method="GET", url=url) as resp:
-					obj = await resp.json()
-					ret = obj['_source']
-					ret['_v'] = obj['_version']
-					ret['_id'] = obj['_id']
-					return ret
+				async with self.session().request(
+					method="GET",
+					url=request_url,
+					ssl=self.SSLcontext,
+					headers=self.Headers,
+				) as resp:
+					if resp.status == 401:
+						raise ConnectionRefusedError("Response code 401: Unauthorized. Provide authorization by specifying either user name and password or api key.")
+					elif resp.status not in {200, 201}:
+						resp = await resp.json()
+						raise ConnectionError("Failed to retrieve data from ElasticSearch. Got {}: {}".format(
+							resp.get("status"),
+							resp.get("error", {}).get("reason")
+						))
+					else:
+						obj = await resp.json()
+						if not obj.get("found"):
+							raise KeyError("No existing object with ID {}".format(obj_id))
+						ret = obj['_source']
+						ret['_v'] = obj['_version']
+						ret['_id'] = obj['_id']
+						return ret
 			except aiohttp.client_exceptions.ClientConnectorError:
-				if url == self.Storage.ServerUrls[-1]:
-					raise Exception("Failed to connect to '{}'".format(url))
+				if url == self.ServerUrls[-1]:
+					raise ConnectionError("Failed to connect to '{}'".format(url))
 				else:
 					L.warning("Failed to connect to '{}', iterating to another cluster node".format(url))
 
+
 	async def get_by(self, collection: str, key: str, value, decrypt=None):
 		raise NotImplementedError("get_by")
 
-	async def delete(self, index, _id=None):
+	async def delete(self, index: str, _id=None) -> dict:
+		"""Delete an entire index or document from that index.
+
+		Args:
+			index: Index to delete.
+			_id: If specified, only document with the ID is deleted.
+
+		Raises:
+			ConnectionRefusedError: Authorization required (status 401)
+			KeyError: No existing object with ID
+			ConnectionError: Unexpected status code
+			Exception: ClientConnectorError
+
+		Returns:
+			The deleted document or message that the entire index was deleted.
+		"""
 		for url in self.ServerUrls:
 			try:
 				if _id:
-					url = "{}{}/_doc/{}?refresh={}".format(url, index, _id, self.Refresh)
+					request_url = "{}{}/_doc/{}?refresh={}".format(url, index, _id, self.Refresh)
 				else:
-					url = "{}{}".format(url, index)
-				async with self.session().request(method="DELETE", url=url) as resp:
-					assert resp.status == 200, "Unexpected response code: {}".format(resp.status)
-					resp = await resp.json()
+					request_url = "{}{}".format(url, index)
+				async with self.session().request(
+					method="DELETE",
+					url=request_url,
+					ssl=self.SSLcontext
+				) as resp:
```

Comment:
Do we need Headers here as well? So that we can delete things with api_key?

Reply:
`				async with self.session().request(
					method="DELETE",
					url=request_url,
					ssl=self.SSLcontext,
					headers=self.Headers,
				) as resp:`